ClangSharp.PInvokeGenerator/PInvokeGenerator.cs

```diff
@@ -3,43 +3,119 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
-using ClangSharp;
+using System.Text;
 
-namespace ClangSharpPInvokeGenerator
+namespace ClangSharp
 {
-    internal sealed class CursorWriter : IDisposable
+    public sealed class PInvokeGenerator : IDisposable
     {
-        private readonly Dictionary<string, OutputBuilder> _outputBuilders;
+        private const int DefaultStreamWriterBufferSize = 1024;
+        private static readonly Encoding DefaultStreamWriterEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        private readonly CXIndex _index;
+        private readonly OutputBuilderFactory _outputBuilderFactory;
+        private readonly Func<string, (Stream stream, bool leaveOpen)> _outputStreamFactory;
         private readonly HashSet<Cursor> _visitedCursors;
-        private readonly ConfigurationOptions _config;
+        private readonly List<Diagnostic> _diagnostics;
+        private readonly PInvokeGeneratorConfiguration _config;
 
         private OutputBuilder _outputBuilder;
         private int _outputBuilderUsers;
+        private bool _disposed;
 
-        public CursorWriter(ConfigurationOptions config)
+        public PInvokeGenerator(PInvokeGeneratorConfiguration config, Func<string, (Stream stream, bool leaveOpen)> outputStreamFactory = null)
```

Comment:
I'm not exactly pleased with the `outputStreamFactory`, but it seems to work. The alternative I came up with was an interface with a single method or something similar....